File: yukicoder/tyama_yukicoder648_isqrt.c
#include <stdio.h>
unsigned long long isqrt(unsigned long long n){
  if(n<=0)return 0;
  if(n<4)return 1;
  unsigned long long x=0,y=n;
  for(;x!=y&&x+1!=y;)x=y,y=(n/y+y)/2;
  return x;
}
int main(){
  unsigned long long n;
  scanf("%llu",&n);
  unsigned long long x=n*8+1,q=isqrt(x);
  if(q*q!=x||q%2==0)puts("NO");
  else printf("YES\n%llu\n",q-1>>1);
}

File: yukicoder/tyama_yukicoder648_isqrt.rb
#!/usr/bin/ruby
class Integer
  def isqrt
    return 0 if self<=0
    return 1 if self<4 # 1
    x,y=0,self
    while x!=y&&x+1!=y
      x,y=y,(self/y+y)/2
    end
    x
  end
end
n=8*gets.to_i+1
q=n.isqrt
if q**2!=n||q%2==0
  puts :NO
else
  puts :YES
  p q-1>>1
end
